### PR TITLE
Broader REVIEW for RFC3 - Index Management

### PR DESCRIPTION
**Pull request by simonbasle**
This PR serves as the medium for broader review of #6 Index Management.
It will transition it from REVIEW to ACCEPTED status, leaving a window for comments until 2016/4/13 (1 week from now).

### CODE REVIEW

**Comment by simonbasle** [comment]
Feedback received during the REVIEW phase:

 - [ ] The `watchIndexes` boolean "`includePrimary`" feels clunky and can always be replaced by the primary name (default "`#primary`") anyway, should probably be removed
 - [ ] With FTS coming, having a differenciating prefix in method names would help (proposed "`n1ql`" as in `createN1qlIndex`, `dropN1qlPrimaryIndex`...)
 - [ ] `IndexInfo` should maybe be optional for languages where value objects are considered more of an hindrance and JSON is a first class citizen (eg. Python, JavaScript...)
 - [ ] The `createN1qlIndex` method should allow for a form of `Expression` to be given in order to specify a `WHERE` clause for the index.

I'll check the items that have been integrated into the RFC via subsequent commit.

**Comment by simonbasle** [comment]
@mnunberg @jeffrymorris I have taken your feedback into account with the above. @jeffrymorris can you please confirm the .NET specifics section, notably, looks good?

**Comment by simonbasle** [comment]
note: commit 7c36bca is the one referred to in previous comment, now rebased on master

**Comment by simonbasle** [comment]
@couchbaselabs/sdk-team it is time for sign off!

Please use the "Add your reaction" feature (in the comment's header toolbar, the smiley face with a + sign) => use the 👍 icon to sign off.

I'll update the document right before squashing and merging.

> PS: if you have implemented the previous state of the RFC (REVIEW) but not the changes that have been made in this PR (ACCEPTED), make sure to open a ticket to track that (and optionally comment with the ticket link here, or in issue #6, or whatever rocks your boat)

**Comment by ingenthr** [comment]
Per meeting on 10 May, @avsej and @mnunberg are going to add the PHP and C implementations to this branch of pull request.

**Comment by simonbasle** [comment]
still need @avsej to provide PHP quick example of how the expected API will look like, @brett19 same for Go.
@couchbaselabs/sdk-team also need your voting on the [comment above](https://github.com/couchbaselabs/sdk-rfcs/pull/35#issuecomment-216960096) for final signoff !!!

**Comment by ingenthr** [comment]
Per 12 May discussion, it looks like we'd like a go sample too.  @brett19 can you add this easily?

- [ ] Need a Go sample from @brett19 
- [ ] @simonbasle going to review the reformatting
- [ ] @avsej need you to fix your signoff, it looks like your commit didn't change the table

File: rfc/0003-indexmanagement.md [hunk 1]
```diff
@@ -63,26 +85,41 @@ String state //eg. "online", "pending" or "deferred"/"building"
 String keyspace
 String namespace
 JsonArray indexKey //or equivalent Json representation
+String condition //empty string if the index has no WHERE clause
```
Comment (jeffrymorris) [inline rfc/0003-indexmanagement.md]: I am not sure what this is? Can we have a better description?

**Comment by jeffrymorris** [comment]
One question regarding the "Condition" field otherwise looks good to me :) and .NET source is updated.

File: rfc/0003-indexmanagement.md (hunk 1 above)
Comment (simonbasle) [inline rfc/0003-indexmanagement.md]: this is the condition part of the index, when you've built it with a `whereClause` expression. The json returned by querying "system:indexes" can contain a `condition` field. If not present, the SDK object should have the empty string instead.

Comment (ingenthr) [inline rfc/0003-indexmanagement.md]: Per discussion 17 May, we don't need to add to the RFC any further description

**Comment by ingenthr** [comment]
Per 19 May review, it looks like the only thing we need to do have you review formatting @simonbasle

**Comment by ingenthr** [comment]
@simonbasle we noted that the recent examples from @brett19 and @jeffrymorris follow a slightly different format in the review on 19 May.  In particular, .NET and go don't show the where clause.  Can you review and come up with whatever we need to try to achieve parity?

**Comment by simonbasle** [comment]
@ingenthr @jeffrymorris in the .NET implementation, the where clause is assembled from lambda expressions given in place of just the names of the fields to index. I think Jeff found it more idiomatic. @jeffrymorris how much of an effort would it be to support both modes (if it doesn't already)?

@brett19, need your input on Go. Do you foresee a problem in providing an (optional) where clause to the index definition?

File: rfc/0003-indexmanagement.md [hunk 2]
```diff
@@ -174,105 +232,245 @@ For SDKs with configurable logging semantics, a mean of toggling a DEBUG log of
...
+     * an explicit T_GSI or T_VIEW.
+     * When using JSON, specify `"using":"gsi"`
+     */
+    unsigned ixtype;
+} lcb_N1INDEXSPEC;
+
+typedef struct {
+    lcb_N1INDEXSPEC spec;
+    lcb_N1IXMGMTCALLBACK callback;
+} lcb_CMDN1IXMGMT;
```
Comment (avsej) [inline rfc/0003-indexmanagement.md]: @mnunberg 

File https://github.com/couchbase/libcouchbase/blob/2.6.0/include/libcouchbase/ixmgmt.h uses **ix** for index in name, so the structure in this spec, but released headers use just X, like `lcb_CMDN1XMGMT`
https://github.com/couchbase/libcouchbase/blob/2.6.0/include/libcouchbase/ixmgmt.h#L138

I think that the spec have to be corrected at least

**Pull request closed without merging**